### PR TITLE
Proxy auth with custom HTTP transport

### PR DESCRIPTION
**Pull request by hhirtz**
### What does this PR try to resolve?

This PR makes cargo authenticate against proxies when needed.
Closes https://github.com/rust-lang/cargo/issues/7330

Add three settings:
- http.proxy-auth
- http.proxy-username
- http.proxy-password

proxy-auth defaults to "auto" so that most use cases work out of the box. I don't know why you would disable authentication support, but I added a "disable" option in this case. You can also force the use of a mechanism and abort if the proxy doesn't advertise it.

proxy-username and proxy-password: `String` vs `Option<String>`?
curl doesn't negotiate (SPNEGO, here "gss") if handle.proxy_username and handle.proxy_password aren't called. This authentication mechanism doesn't require a user/pass in cargo config. This shouldn't hurt (?) when the proxy doesn't require authentication.

About curl-rust's spnego and ntlm features:
https://github.com/alexcrichton/curl-rust/blob/b50c5d355aab69483752db51815c5a679b29d6c4/curl-sys/Cargo.toml#L54-L60
**This is the last blocker I think.** If we require those features, then vendored curl builds will require "a suitable GSS-API library or SSPI on Windows" [0] for the "gss" setting to work (which is what corporate proxies are using most of the time). I think this PR is blocked until we can either  vendor those dependencies, or we change the official cargo builds to link dynamically to libcurl.

The official cargo builds seems to be made with vendored curl.

### How should we test and review this PR?

Complete testing is a bit tedious since you'd need e.g. squid [1] to test each authentication mechanism. With such proxy, run `cargo update` with http-proxy environment variables (plus `--config http.proxy-{username,password}=` if needed).

[0] https://docs.rs/curl/0.4/curl/easy/struct.Auth.html#method.gssnegotiate
[1] http://www.squid-cache.org/

### CODE REVIEW

File: src/cargo/util/config/mod.rs [hunk 1]
```diff
@@ -2215,10 +2215,43 @@ impl Drop for PackageCacheLock<'_> {
...
 #[derive(Debug, Default, Deserialize, PartialEq)]
 #[serde(rename_all = "kebab-case")]
 pub struct CargoHttpConfig {
     pub proxy: Option<String>,
+    #[serde(default)]
+    pub proxy_auth: CargoHttpProxyAuth,
+    #[serde(default)]
+    pub proxy_username: String,
+    #[serde(default)]
+    pub proxy_password: String,
```
Comment (arlosi) [inline src/cargo/util/config/mod.rs:2254]: I'd prefer the username and password be `Option`s rather than empty strings.

```suggestion
    pub proxy_username: Option<String>,
    pub proxy_password: Option<String>,
```

File: src/cargo/util/config/mod.rs [hunk 2]
```diff
@@ -2215,10 +2215,43 @@ impl Drop for PackageCacheLock<'_> {
...
+    Basic,
+    Digest,
+    Gss,
+    Ntlm,
+}
+
+impl CargoHttpProxyAuth {
+    pub fn to_easy(&self) -> Auth {
+        let mut easy = Auth::new();
+        let easy = match self {
```
Comment (arlosi) [inline src/cargo/util/config/mod.rs:2233]: Avoid the `easy.clone()` below by not redefining `easy` here. Also consider calling it `auth` instead of `easy`
```suggestion
        match self {
```

File: src/cargo/util/config/mod.rs [hunk 3]
```diff
@@ -2215,10 +2215,43 @@ impl Drop for PackageCacheLock<'_> {
     }
 }
 
+#[derive(Debug, Default, Deserialize, PartialEq)]
+#[serde(rename_all = "kebab-case")]
+pub enum CargoHttpProxyAuth {
+    #[default]
+    Auto,
```
Comment (arlosi) [inline src/cargo/util/config/mod.rs:2222]: Does the default of `Auto` introduce any additional HTTP requests for the non-proxy case?

The [curl docs](https://curl.se/libcurl/c/CURLOPT_PROXYAUTH.html) indicate that it might:
> If more than one bit is set, libcurl will first query the site to see what authentication methods it supports and then pick the best one you allow it to use. For some methods, this will induce an extra network round-trip

File: src/cargo/util/config/mod.rs (hunk 2 above)
Comment (hhirtz) [reply src/cargo/util/config/mod.rs:2233]: It's not possible to avoid the clone because `match self { .. }` is a `&mut Auth`. Also `Auth` does not implement `Copy`.

https://docs.rs/curl/0.4.44/curl/easy/struct.Auth.html

File: src/cargo/util/config/mod.rs (hunk 3 above)
Comment (hhirtz) [reply src/cargo/util/config/mod.rs:2222]: This change focuses on proxy authentication specifically, not authentication to arbitrary HTTP servers. The `Auth` API is used for both problems, thus this paragraph I think. In the non-proxy case there is no additional request to be made.

This could well send an additional request to a proxy that doesn't need authentication, however I don't think it would make a noticeable difference in performance compared to, e.g. the time it takes to update the registry or download kilobytes of crates (?).

File: src/cargo/util/config/mod.rs (hunk 2 above)
Comment (arlosi) [reply src/cargo/util/config/mod.rs:2233]: You don't need to capture the result of the `match` in a variable. The `basic()`, `digest()`, etc methods mutate the `Auth` struct, so you can return the one you created with `Auth::new()` without cloning.
```
    pub fn to_easy(&self) -> Auth {
        let mut auth = Auth::new();
        match self {
            Self::Auto => auth.basic(true).digest(true).gssnegotiate(true).ntlm(true),
            Self::Disable => &auth,
            Self::Basic => auth.basic(true),
            Self::Digest => auth.digest(true),
            Self::Gss => auth.gssnegotiate(true),
            Self::Ntlm => auth.ntlm(true),
        };
        auth
    }
```

File: src/cargo/ops/registry.rs [hunk 4]
```diff
@@ -560,6 +560,9 @@ pub fn configure_http_handle(config: &Config, handle: &mut Easy) -> CargoResult<
     if let Some(proxy) = http_proxy(config)? {
         handle.proxy(&proxy)?;
     }
+    handle.proxy_auth(&http.proxy_auth.to_easy())?;
```
Comment (arlosi) [inline src/cargo/ops/registry.rs:563]: If this is `CargoHttpProxyAuth::Disable` should we avoid calling any of the `proxy_` methods?

File: src/cargo/ops/registry.rs [hunk 5]
```diff
@@ -560,6 +560,9 @@ pub fn configure_http_handle(config: &Config, handle: &mut Easy) -> CargoResult<
...
+    handle.proxy_username(http.proxy_username.as_deref().unwrap_or(""))?;
```
Comment (arlosi) [inline src/cargo/ops/registry.rs:564]: Can you add a comment about why these need to be called with empty string (instead of just not calling them)?

Comment (hhirtz) [reply src/cargo/ops/registry.rs:564]: The comment is in the commit message [0] (second paragraph). Do you want another one in the code?

[0] <https://github.com/rust-lang/cargo/pull/11433/commits/e7a590516c58b85dfaa98bfcdfb7f77c9b551173>

**Comment by ehuss** [comment]
Unfortunately this seems like a relatively meaty change, so it can be difficult to nail everything down.

@hhirtz Is there a specific authentication scheme that you personally are interested in using? I was wondering if there is some way to narrow the scope of what is being added here. For example, if cargo needs to move to a different backend, then it would need to support all the mechanisms added here, which can be a large burden.  Additionally, it looks like building curl with all the appropriate support looks potentially complicated. Exactly what will be supported on all of Cargo's platforms with the current build system?

Another issue is that we try to avoid including secrets in the configuration files. Adding the password here seems like an anti-pattern, so I think we may need to look at some alternate approach. I'm not sure if that involves using `credentials.toml` or something else.

In general, it can also be difficult for me to review and understand a change like this since I am not deeply familiar with all the different proxy authentication mechanisms. I think to move forward, I would need some kind of ELI5 introduction to what is being added here to help evaluate how these will be used and the consequences.
